### PR TITLE
Feature/ds and location

### PR DESCRIPTION
**Pull request by dsmith42**
Refactored the Android location PR onto Fragment FindMeal.

Again, please take a look and if there are better ways to approach this in a more Android and clean manner, please do let me know.

Going to refactor into ViewModel next.

### CODE REVIEW

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt [hunk 1]
```diff
@@ -28,48 +41,161 @@ class FindMealFragment : Fragment() {
     private lateinit var swipeRefreshLayout: SwipeRefreshLayout
 
     private val sdk = MealsSDK()
-    private val mealsAdapter = MealsAdapter(listOf())
+    private var distanceUnit: DistanceUnit = DistanceUnit.miles
+    private val mealsAdapter = MealsAdapter(listOf(), distanceUnit)
+
+    private lateinit var fusedLocationClient: FusedLocationProviderClient
+    private val permissionResultParser: PermissionResultParser by lazy {
```
Comment (KaneCheshire) [inline lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt]: Doesn't need to be lazy

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/adapter/MealsAdapter.kt [hunk 2]
```diff
@@ -52,7 +53,7 @@ class MealsAdapter(var mealsList: List<Meal>) :
             }
 
             // TODO calculate distance (this could be done in the shared layer)
```
Comment (Gmiller290488) [inline lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/adapter/MealsAdapter.kt]: Can we get rid of the TODO now?

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt [hunk 3]
```diff
@@ -28,48 +41,161 @@ class FindMealFragment : Fragment() {
...
+        PermissionResultParser()
+    }
+    private var lastLocation: Location? = null
+    private var hasSetupLocationUpdates = false
```
Comment (KaneCheshire) [inline lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt:52]: When this is in a VM lets see if we can do away with some of these states

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/adapter/MealsAdapter.kt [hunk 4]
```diff
@@ -52,7 +53,7 @@ class MealsAdapter(var mealsList: List<Meal>) :
...
-            distanceView.text = "8637.35km"
+            distanceView.text = "${meal.distance} ${distanceUnit}"
 
             // TODO format these dates
```
Comment (Gmiller290488) [inline lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/adapter/MealsAdapter.kt]: I'm currently working on formatting dates in the shared area also so can remove this :-)

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt [hunk 5]
```diff
@@ -28,48 +41,161 @@ class FindMealFragment : Fragment() {
...
+    private var locationUpdateState = false
+
+    private val locationCallback = object : LocationCallback() {
+        override fun onLocationResult(locationResult: LocationResult?) {
+            lastLocation = locationResult?.lastLocation ?: return
+        }
+    }
 
     companion object {
+        val LOCATION_PERMISSION_REQUEST_CODE = 1
```
Comment (KaneCheshire) [inline lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt]: These can probably be `private` and as a small thing I'd move to either the top or bottom of the class

**Review by KaneCheshire — APPROVED** [review]

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt [hunk 6]
```diff
@@ -28,48 +41,161 @@ class FindMealFragment : Fragment() {
...
+
+        task.addOnSuccessListener {
+            locationUpdateState = true
+        }
+        task.addOnFailureListener { e ->
+           if (e is ResolvableApiException) {
+               try {
+                   e.startResolutionForResult(this.requireActivity(), REQUEST_CHECK_SETTINGS)
+               } catch (sendEx: IntentSender.SendIntentException) {
+                   //
```
Comment (Gmiller290488) [inline lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt]: something missing?

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt [hunk 7]
```diff
@@ -28,48 +41,161 @@ class FindMealFragment : Fragment() {
...
+           }
+
+        }
+        hasSetupLocationUpdates = true
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+        if (requestCode == REQUEST_CHECK_SETTINGS) {
+            if (resultCode == Activity.RESULT_OK) {
```
Comment (Gmiller290488) [inline lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt]: could it be combined?

File: shared/src/commonMain/kotlin/com/kcc/kmmhackathon/shared/MealsSDK.kt [hunk 8]
```diff
@@ -18,10 +18,6 @@ class MealsSDK {
         return api.getAllMeals()
     }
 
-    @Throws(Exception::class) suspend fun getMeals(userLat: Double, userLon: Double, distanceUnit: DistanceUnit, forceReload: Boolean): List<MealWithDistance> {
-        return api.getAllMeals(userLat, userLon, distanceUnit)
-    }
-
```
Comment (Gmiller290488) [inline shared/src/commonMain/kotlin/com/kcc/kmmhackathon/shared/MealsSDK.kt:24]: 🎉

**Review by Gmiller290488 — COMMENTED** [review]
Come you also remove the `MealWithDistance` data class from Entity.kt in this PR please (if it's not being used anywhere else on Android?)

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/adapter/MealsAdapter.kt (hunk 2 above)
Comment (dsmith42) [reply lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/adapter/MealsAdapter.kt]: Consider it gone

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt (hunk 3 above)
Comment (dsmith42) [reply lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt:52]: Sounds good

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt (hunk 6 above)
Comment (dsmith42) [reply lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt]: Yes, have updated the catch now to Log exception out

File: lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt (hunk 7 above)
Comment (dsmith42) [reply lib-find-meal/src/main/java/com/kinandcarta/lib/find/meal/view/FindMealFragment.kt]: Nice have updated to reflect suggestion

**Review by Gmiller290488 — APPROVED** [review]

**Pull request merged**